src/observer.tsx

```diff
@@ -40,7 +40,7 @@ export function observer<T extends IReactComponent>(component: T): T {
             throw new Error("render property of ForwardRef was not a function")
         return React.forwardRef(function ObserverForwardRef() {
             return <Observer>{() => baseRender.apply(undefined, arguments)}</Observer>
-        }) as any
+        }) as T
```

Comment:
I am unsure of the casting and wonder if there is a better way.  Open to ideas.

Reply:
Not sure either, I would be expecting it would work directly as those types are compatible. Unless that generic argument for props is a problem somehow. The `React.forwardRef<R, P>` does probably generate the same output.  Perhaps you can try this and then utilize the `P` there 🤷‍♂ 

`observer<P extends object, T extends IReactComponent<P>>(component: T): T {`

Reply:
I think for observer we should keep exactly the same signature as we had before in the `.d.ts`, two not be confused with two changes at the same time.... That means the old TS types should remain working out of the box?

Reply:
Well, I think it should be just fine. Since the `P` is referenced in the `T`, it won't be required to be used. I highly doubt anyone has been supplying `T` with previous versions.

https://www.typescriptlang.org/play/index.html#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wChSkAPSWOGATzCTgEll0YBhXSAOyV4wAPAAU4AXjgpe9AHwTScJXAA+iYjAB0nADYoAzvuBpu4CP0G6D+0bMXK17DNp7mBXPYdv2ljjZoBlGBQYJB0kQ1M+d29lVXUOTWRMADFoAHcUKAATYF4Acyi3QW9STABXXgxgczgIACN9JCgAN2bROCpQ3mz9OvqAKyQMABo4ABVOym7e1icuVwthEVlZAAo0RfcALgmASl3JgG8fOCIYcqheOE2zJdIAX3IGptbmtbW9iXlecp0dPZAA

On the contrary, with this updated variant we are aligning with `observer` from `mobx-react-lite` which does accept `P` as generic arg.

Reply:
Yeah, I tried your suggestion and I still get the same error.  Where required to cast to T the return value of  `forwardRef` (ExoticComponent), `observerLite` (FunctionComponent), and `makeClassComponentObserver` (ComponentClass).  In order to avoid

`Type is not assignable to type 'T'.
  Type is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'IReactComponent<P>'`

Reply:
Well, and did you try `React.forwardRef<any, P>`?

Reply:
Yes, I believe so.  Like this?

`observer<P extends object, T extends IReactComponent<P>>(component: T): T`

`React.forwardRef<any, P>(function ObserverForwardRef() {`

Output:
`Type 'ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<any>>' is not assignable to type 'T' `

Reply:
I guess all in all I'm thinking we keep the signature the same for now so that there are no regressions.  Then, as I work through the tests, I will also spend some time coming up with a better solution (if any exists).  I wouldn't want this to hold up a potential release.

Reply:
Alright, fair enough, let's keep it simple for now and feel free to look into it over the time.